core/chains/evm/logpoller/parser.go

```diff
@@ -0,0 +1,502 @@
+package logpoller
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+const (
+	sequenceFieldAlias = "sequence"
+	sequenceField      = "CONCAT(block_number, '-', tx_hash, '-', log_index)"
+	blockFieldName     = "block_number"
+	timestampFieldName = "block_timestamp"
+	txHashFieldName    = "tx_hash"
+	eventSigFieldName  = "event_sig"
+	chainIDFieldName   = "evm_chain_id"
+)
+
+var (
+	ErrUnexpectedCursorFormat = errors.New("unexpected cursor format")
+)
+
+// pgDSLParser is a visitor that builds a postgres query and arguments from a commontypes.QueryFilter
+type pgDSLParser struct {
+	args *queryArgs
+
+	// transient properties expected to be set and reset with every expression
+	expression string
+	err        error
+}
+
+var _ primitives.Visitor = (*pgDSLParser)(nil)
+
+func (v *pgDSLParser) Comparator(p primitives.Comparator) {}
+
+func (v *pgDSLParser) Block(p primitives.Block) {
+	cmp, err := cmpOpToString(p.Operator)
+	if err != nil {
+		v.err = err
+
+		return
+	}
+
+	v.expression = fmt.Sprintf(
+		"%s %s :%s",
+		blockFieldName,
+		cmp,
+		v.args.withIndexedField(blockFieldName, p.Block),
+	)
+}
+
+func (v *pgDSLParser) Confirmations(p primitives.Confirmations) {
+	switch p.ConfirmationLevel {
+	case primitives.Finalized:
+		v.expression = v.nestedConfQuery(true, 0)
+	case primitives.Unconfirmed:
+		p.ConfirmationLevel = primitives.ConfirmationLevel(0)
+
+		fallthrough
+	default:
+		confs := evmtypes.Confirmations(p.ConfirmationLevel)
+
+		if p.ConfirmationLevel == primitives.Unconfirmed {
+			confs = evmtypes.Unconfirmed
+		}
+
+		v.expression = v.nestedConfQuery(false, uint64(confs))
+	}
+}
```

Comment:
Does this do the same thing?

```
func (v *pgDSLParser) Confirmations(p primitives.Confirmations) {
	switch p.ConfirmationLevel {
	case primitives.Finalized:
		v.expression = v.nestedConfQuery(true, 0)
		v.expression = nestedBlockNumberQuery(evmtypes.Finalized)
	case primitives.Unconfirmed:
		v.expression = nestedBlockNumberQuery(evmtypes.Unconfirmed)
	default:
		v.expression = nestedBlockNumberQuery(evmtypes.Confirmations(p.ConfirmationLevel))
	}
}
```

Reply:
Yes. Finalized is a special subquery, separate from Unconfirmed and specific confirmation values. Unconfirmed is just an alias for the case of 0 confirmations. That is why the fallthrough exists. I can clean it up a bit though.